Win32 test std

coreyapowell opened the pull request:
tested in Win11 MSYS2 Mingw64, Clang64, WSL Debian, MSVC 2023, Clang in MSVC 2022
finished modifications so that test_std can pass audit in windows. 
adds Win32 os_symlink
modifies std_tmpfile for MINGW
adds destructor to std_File for tmpfile deletion

this does not contain os.exec or inclusion of Win32 std_test.js for that. It didn't test windows before, and that can probably be moved from test_std (considering it's not std it's os anyway). my change is separated and that test script is separated in that change.
test_std no longer tries to create  a symlink in windows because you need admin rights, but it works if you test it manually. 
also there's a time bug returning negative number I didn't get into removed from test_std

coreyapowell committed:
Add files via upload

coreyapowell committed:
Add files via upload

coreyapowell committed:
Add files via upload

coreyapowell committed:
Add files via upload

coreyapowell committed:
Update test_std.js

I put this back like it was. 
Changes will be split into smaller parts and I've gotten confused when permissions went back and forth

coreyapowell committed:
Merge pull request #2 from coreyapowell/backto_original

Update test_std.js

coreyapowell committed:
Delete test_std.js

getting back to original codebase

coreyapowell committed:
Update quickjs-libc.c

modifies std_tmpfile to work properly with MINGW
(I think MSVC will work okay)
std_file finalizer modified
creates os_symlink for win32

coreyapowell committed:
Update test_std.js

moved a time issue out of WIN32 section. 
I moved os.symlink out as well and commented, you need admin rights or a setting to execute. test works now with these changes in Windows 11.

coreyapowell commented:
I keep seeing "we're all done". Great! I don't know if I'm supposed to leave this branch in place or merge it in my main.

bnoordhuis reviewed (commented):
Same comment as the other PR, can you fix the style issues first? Thanks.

coreyapowell commented:
Sure! I'll also pull out all of the comments and flags. I'm trying to understand how you guys collaborate as independent users.  I want whoever is handling the windows portion to at least have a heads up. It's as 'drop in replacement' as I could manage. I just expected someone might have some feedback.

I don't want to trample over everyone else's projects. In fact, I believe Mr Bellard likely ran into the same issues (that windows doesn't play nice with FDs and PIDs) and just skipped all of this. The quickjs-ng team added worker pipes for win32 (in quickjs-libc.c), but they use HND types in a piece of code that users won't see in the script. The quickjs documentation, the current std and os module, the c standard, and all of linux use and prefer PIDs and FDs. I haven't had a chance to test things like passing the FD down a pipe to another .exe or to a worker thread. Passing out HNDs in script would guarantee that flexibility, but it would be a documentation nightmare. So this works, it does what I need it to do as a scripting tool, and it function simply enough. 

I've created one replacement function that can be used in linux and windows without testing the environment, but the user needs to understand what that function does. Windows doesn't have waitpid. Do I just leave it in the tests.js folder and let them figure it out? Leave my comments above the c code and test.js code? Or put a comment in the documentation section for watchpid next to waitpid? Is everyone okay with the way I split those functions up in test_os_exec.js? I am just trying to make the use clear. Is everyone happy with the way I've left that? 

I appreciate the feedback so far, I don't expect anyone to take up all their time on this. I just know what any windows user or admin needs, and I'm trying to complete it.

coreyapowell committed:
took out some comments

 ... okay, sorry, I took out more than comments. I had some printf functions left in there! I think that warrants some testing, that I didn't see that printf!
The reason I didn't see that printf was that windows 10 requires admin rights to test symlinks, and I have it disabled in the test script.
I've left a documented flag in test_std.js and properly tested this again. MING64, CLANG64, MSVC, ClangCL with VC Build

coreyapowell committed:
fixed comments and proper symlink test flag

 const IS_WIN_ADMIN_TEST_FLAG = 0;
modified so that someone can run as admin with this flag on and actually test symlink.
comment above that clarified that "this code is broken".  - That's not my code, my code makes the majority of the std_file finish executing. I just can't readlink from windows , and I can't do that mtime.
...... (but I don't need metric time, I'm not a scientist!

coreyapowell committed:
Update quickjs-libc.c

*space*

coreyapowell committed:
Update quickjs-libc.c

...

coreyapowell commented:
I didn't think I'd get to it tonight, but I've cleaned these files up.  (just the tmpfile/symlink files)
I had accidentally posted some comments in the wrong thread earlier, if you got a message about the other code.

 I don't know if you'd seen this, I had printf's still in the code for symlink.. 
You have to be an administrator in windows 10+ now, so I was manually moving that around in the test scripts. I put a flag in test_std.js that can be changed and execute the code. I'm glad I did because going back and forth between these environments and running all these compilers and test scripts things like that get missed! I don't know if you guys have a one-stop place to execute the tests, I'm manually executing the scripts in the test folder. And not all of them work.

That was the whole point of even fixing symlink in windows. I don't need it! I just wanted to see the test scripts complete. I might use tmpfile more.

Also, and I'm asking this. maybe I should post a bug discussion? The line before symlink, testing mtime, reports -1000 and expects 10000 or something on all 4 or my windows builds. That's broken too, and that's not something I immediately knew how to fix. So I've bypassed it in the test but it used to be there. I can get to it in the future if it's fixable, but I need this functionality updated so that I can use Quickjs to run my compiles and detect error codes and link my modules for me. I'm doing things the hard way for now!

saghul commented on quickjs-libc.c:
```diff
@@ -1034,6 +1034,14 @@ typedef struct {
+typedef struct {
+    FILE *f;
+    int is_kind; 
+    char filename[64];
```
Maybe use JS__PATH_MAX to be safe?

saghul commented on quickjs-libc.c:
```diff
@@ -1034,6 +1034,14 @@ typedef struct {
+#if defined(__MINGW32__) || defined(__MINGW64__)
+typedef struct {
+    FILE *f;
+    int is_kind; 
```
Use an enum to be more readable please.

saghul commented on quickjs-libc.c:
```diff
@@ -1207,6 +1224,63 @@ static JSValue js_std_fdopen(JSContext *ctx, JSValueConst this_val,
+        env = getenv("TEMP");
+    int i = 0;
+    if (env) {
+        while (env[i]) {
```
Make the buffer long enough or malloc it, this feels wrong.

saghul commented on quickjs-libc.c:
```diff
@@ -1207,6 +1224,63 @@ static JSValue js_std_fdopen(JSContext *ctx, JSValueConst this_val,
+    }
+    char* fname = &s->filename[i];
+    char* templ = "\\qXXXXXXX";
+    while (templ[0]) {
```
memcpy?

saghul commented on quickjs-libc.c:
```diff
@@ -1207,6 +1224,63 @@ static JSValue js_std_fdopen(JSContext *ctx, JSValueConst this_val,
 
 #if !defined(__wasi__)
+#if defined(__MINGW32__) || defined(__MINGW64__)
+static JSValue js_std_tmpfile(JSContext *ctx, JSValueConst this_val,
```
Actually, why are we doing all of this when MinGW already has `tmpfile` ?

coreyapowell replied:
because it has been broken for years. it does not work. Windows permissions changed and you cannot write to the same place that they used to write to.
It's hit and miss with microsoft compilers, because the internal windows code that MINGW calls is still there and it doesn't work. I researched it. It's a folder permission change in the operating system, in at least back to win10.

saghul replied:
Is this the problem? https://github.com/msys2/MINGW-packages/issues/18878#issuecomment-2254372767

Perhaps it's time to default to the UCRT build.

saghul commented on tests/test_std.js:
```diff
@@ -298,3 +317,5 @@ test_interval();
 test_timeout_order();
 test_stdio_close();
+
+std.exit(0);
```
Drop this.

saghul commented on quickjs-libc.c:
```diff
@@ -3072,12 +3146,25 @@ static JSValue js_os_symlink(JSContext *ctx, JSValueConst this_val,
         return JS_EXCEPTION;
     }
+#ifdef _WIN32
+    int isdirflag = 0; 
```
What is this used for?

coreyapowell replied:
Well you need to tell me if you want this. it is a flag that can be passed if  you want to create a directory link.
The line below it was supposed to be: err =  CreateSymbolicLinkA(linkpath, target, ( isdirflag || 2 ) ) ; 
but I thought it was a flag. I took it out. I tested that manually before and didn't add folder symlink testing to test_std.js 
To be honest it should also have a comment: 1 = SYMBOLIC_LINK_FLAG_DIRECTORY if it's left in there.

coreyapowell replied:
if you want the symlink taken out, it's not exactly the same and error prone in windows. I don't need it. It's more of a shortcut I think, and it doesn't warn you if the path isn't real or anything. You have to be an admin to use it. I just wanted to run test_std.js and get it to finish.

coreyapowell replied:
it's a shortcut, and fopen doesn't even work on it. I don't know if it's worth it.

coreyapowell replied:
You've left the change request. Is that for this? 
I programmed it as an optional parameter, so if a user doesn't specify the flag parameter, it then it still creates a file link by default. I will actually use this, for instance, when I download the current release and link test262 (rather than copy 2GB). Windows users are used to this restriction honestly. To me, qjs could be a first class admin tool.

saghul commented on quickjs-libc.c:
```diff
@@ -3072,12 +3146,25 @@ static JSValue js_os_symlink(JSContext *ctx, JSValueConst this_val,
+    if (argc > 2) {
+         if (JS_ToInt32(ctx, &isdirflag, argv[2])) return JS_EXCEPTION;
+    }
+    // 2 = SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE
```
Use the constant rather than the value?

coreyapowell replied:
I had an error that it couldn't find  header and just left it in the comment.
I've reinstalled my Windows LLVM setup since then, the error may have gone away. Lots of bugs have gone away with Clang 20. I'm making sure now that it's there. That version was built when there was way less support, back when V8 added Clang as a build environment. That was before you could just add Clang from the Visual Studio menu.

saghul reviewed (changes requested):
Left some comments, PTAL.

coreyapowell committed:
Update test_std.js

coreyapowell committed:
Update quickjs-libc.c

altercation:
MINGW64 ... no longer MINGW32
no longer duping the tmpfile handle. that was because of some other function and it's not needed.
improved os_tmpfile names of variables for clarity. 
JSTMPFile stores a js_mallocz'd buffer instead. frees buffer.
enumerated JSSTDFileKind kind;

coreyapowell committed:
Update quickjs-libc.c

fixed flag: SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE
tested

coreyapowell commented:
thank  you for the opportunity of getting this in here. changes made.

enummbed JSTMPFileKind. (it's just JSTMPFile, I didn't change JSSTDFile. I don't know who's project that affects. But the bool should be 1 for a pipe in the old code. It should work.
fixed the symbolic_link flag and restored the isdirflag |  symbolic_link
get tmp folder size and malloc filename. 
stores filename as a js_mallocz buffer in JSTMPFile.
I no longer dup the tmpfile handle. I just fopen the FD mkstemp gives me.
clarified the variables a bit.
no longer testing std.exit()

so the best solutions to the symlink dir issue online are to open the raw shortcut or use kernel level functions to read the harddrive node directly. And I'm not getting paid for any of this, so that's not free, my laptop's  not insured. If you want to learn the file format of a shortcut in windows, I'm sure that will work until windows 12 comes along.

coreyapowell committed:
Update test_std.js - no lstat

I have to manually test this change unless we add a detect user or detect group function. Win32 has no lstat.
Moved the comment and admin flag to the top for ease of use until someone comes up with a better solution.

saghul commented on quickjs-libc.c:
```diff
@@ -1044,13 +1053,23 @@ static void js_std_file_finalizer(JSRuntime *rt, JSValueConst val)
+            if (ss->f) fclose(ss->f);
+            if (ss->filename != 0) remove(ss->filename);
+            js_free_rt(rt, ss->filename);
+        } else if (s->f && !is_stdio(s->f)) {
```
I think you can just return here rather than having the check duplicated

coreyapowell replied:
Do you mean return and take the else out?

saghul replied:
Pretty much. Well, you may need some extra cleanup, or a goto end...

coreyapowell replied:
yes I left it open for cleanup. added 'return;' and took out else. I just wasn't sure what you meant.

saghul commented on quickjs-libc.c:
```diff
@@ -1207,6 +1226,62 @@ static JSValue js_std_fdopen(JSContext *ctx, JSValueConst this_val,
+        JS_FreeValue(ctx, obj);
+        return JS_EXCEPTION;
+    }
+    fn_template = "\\qXXXXXXX";
```
Make this static?

coreyapowell replied:
made static. I always assumed the compiler was smart enough to do that. So I researched it and I get it, someone might want their strings on their stack, so the default is not.
I have always seen it in the string table, not on the stack. Researching it, maybe my compilers were optimizing that way, and I just assumed it was static by default. Maybe I'm thinking of a const string.

saghul commented on quickjs-libc.c:
```diff
@@ -1207,6 +1226,62 @@ static JSValue js_std_fdopen(JSContext *ctx, JSValueConst this_val,
+        env_tmp = getenv("TEMP");
+    if (env_tmp) {
+        fn_len = strlen(env_tmp);
+        fn_buff = js_malloc(ctx, fn_len + 10);
```
don't use 10, if that is the length of the template, use strlen

saghul commented on quickjs-libc.c:
```diff
@@ -1207,6 +1226,62 @@ static JSValue js_std_fdopen(JSContext *ctx, JSValueConst this_val,
+        memcpy(fn_buff, env_tmp, fn_len);
+        memcpy(fn_buff + fn_len, fn_template, 10);
+    } else {
+        fn_buff = js_mallocz(ctx, 10);
```
You can share the allocation and copy code across both branches by checking if you have an env var or not.

coreyapowell replied:
It's wrong anyway. Even if there's no %TEMP% I put "/" and that would default to c:\. That defeats the whole point. There are no create permissions there. I should throw an exception if I don't have TEMP. Otherwise I should have added "\\" manually. 
Let me know if you disagree. My original logic was to create in CWD if someone's stripped the environment.

coreyapowell replied:
There's no way around two ifs that I could see. 
I admit, may be clearer for someone else to read the way you're suggesting.

saghul commented on quickjs-libc.c:
```diff
@@ -1207,6 +1226,62 @@ static JSValue js_std_fdopen(JSContext *ctx, JSValueConst this_val,
+    };
+    //int fd = dup(mkf);
+    s->f = fdopen( mk_fd, "a+");
+    if (s->f == 0) {
```
Check `if (!s->f`

saghul commented on quickjs-libc.c:
```diff
@@ -1207,6 +1226,62 @@ static JSValue js_std_fdopen(JSContext *ctx, JSValueConst this_val,
+    //int fd = dup(mkf);
+    s->f = fdopen( mk_fd, "a+");
+    if (s->f == 0) {
+        JS_ThrowInternalError(ctx, "tmpfile failed to open file with error: %d.", errno);
```
You are leaking mk_fd here, you need to close it.

coreyapowell replied:
Yes, I missed that. This got involved. I tried several fixes and found myself going in circles, and I'll tell you why: 
I don't manually handle closing this file. Giving the user the FILE object was just confusing me to no end. But I get it now, once the FILE is in the managed structure, ownership is passed to the finalizer. If they had receive an int for a descriptor, closing the file is their problem. When they receive a FILE, Javascript handles that for them with the finalizer. Otherwise, should we be detecting multiple closes or is it safe to have redundant closes? It's not safe to close descriptors twice. Descriptors are reused. 
Also, all of the examples of the temp file code use dup on that descriptor and then close it. I see no reason to dup that descriptor. 
But I understand you, that the throw should close it, and I missed it. I just want to make sure I'm not missing something in my logic. I'm probably overthinking the problem.

saghul commented on quickjs-libc.c:
```diff
@@ -1207,6 +1226,62 @@ static JSValue js_std_fdopen(JSContext *ctx, JSValueConst this_val,
+        JS_ThrowInternalError(ctx, "tmpfile failed to create file with error: %d.", errno);
+        goto file_failure;
+    };
+    //int fd = dup(mkf);
```
Drop the comment.

saghul commented on quickjs-libc.c:
```diff
@@ -3072,12 +3147,25 @@ static JSValue js_os_symlink(JSContext *ctx, JSValueConst this_val,
+    }
+    err =  CreateSymbolicLinkA(linkpath, target, ( isdirflag | 
+        SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE ) ) ; 
+    if (!err) err = GetLastError();
```
style: please use similar style to what we use all across the codebase

saghul commented on tests/test_std.js:
```diff
@@ -5,6 +5,10 @@ import { assert } from  "./assert.js";
+/* symlink in windows 10+ requres admin or SeCreateSymbolicLinkPrivilege privilege found under:
+ Computer Configuration\Windows Settings\Security Settings\Local Policies\User Rights Assignment\
+ set IS_WIN_ADMIN_TEST_FLAG to 1 and run as Administrator to test win32 os.symlink */
+const IS_WIN_ADMIN_TEST_FLAG = 0;
```
Rather than do this, check the error code and if it's a permission error then don't fail the test.

coreyapowell replied:
removed. 
as per other comment, it's making less sense that windows has done this to the feature.

saghul commented on tests/test_std.js:
```diff
@@ -168,9 +172,12 @@ function test_os()
     assert(st.mode & os.S_IFMT, os.S_IFREG);
-    assert(st.mtime, fdate);
+
+    if (!isWin) // returns some negative value in windows 11. had to remove this on my build. (CAP)
```
Please undo this unrelated change, our test suite does not reproduce the problem.

coreyapowell replied:
I had a minute and figured it out. It just glitches out to set that time before 5am here in NY time. That's 18 million milliseconds. The script needs more than 10,000.
I put 86 million 400 thousand milliseconds, which is 24 hours, which gets any user through Dec 31 1969 to Jan 1.

saghul commented on tests/test_std.js:
```diff
@@ -183,6 +190,14 @@ function test_os()
 
         assert(os.remove(link_path) === 0);
+        
+    } else if (IS_WIN_ADMIN_TEST_FLAG) {
```
As mentioned above, handle the failure gracefully on Windows.

coreyapowell replied:
So we start the test and skip the test? I'll do that, but would you rather remove it at this point?
I know I said I'd use the feature, but now I'm using hard links. I will be frank with you, lstat needs to be working for all of this and that's too large of a change for now.

saghul replied:
Fine by me, we can handle one change at a time. Keep this PR to tmpfile alone then?

coreyapowell replied:
I *just* exit gracefully or remove the link. I suppose you need to test removing the link.

coreyapowell replied:
sorry you commented as I had the screen un-refreshed. It looked ugly to me when I commented but it's growing on me. Silence is golden. It's whatever you want. 
If someone is editing the file to test it, they can put a console.log in to see it. I removed the comment, and I just hate for another windows user to learn this the hard way.
Coincidentally this isn't a problem with hard links. I'm doing hard links in my batch file when I download the git and move test262 and it doesn't complain. If os.exec is fixed, I would just use that. I'm pretty sure that symlink is a soft link isn't it? That kind of makes Microsoft's decision *silly*.

saghul commented on tests/test_std.js:
```diff
@@ -298,3 +313,4 @@ test_interval();
 test_timeout();
 test_timeout_order();
 test_stdio_close();
+
```
Drop the extra line.

coreyapowell replied:
changes in as requested and tested.
I will download and make sure after I've had a break.

coreyapowell committed:
not admin in win32 just skip symlink tests.

removed extra line.
win32 symlink fails on 1314 because admin rights required.
removed flag.

coreyapowell committed:
tested and cleaned up

cleaned up js_os_tmpfile
cleaned up finalizer

coreyapowell committed:
formatting

fixed indention level

coreyapowell commented:
requested changes made.
downloaded from git and executed tests.
executed test262 on debian.

I've updated the merged copy of these two changes and I work from that, for my stuff. And I'm about a step away from having the windows version running test262. I used your cmake but I did manage to get Mingw64 to run it.
